flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java

```diff
@@ -31,28 +34,37 @@
 public class PluginConfig {
 	private final Optional<Path> pluginsPath;
 
-	private PluginConfig() {
-		this.pluginsPath = Optional.empty();
-	}
+	private final String[] alwaysParentFirstPatterns;
 
-	private PluginConfig(Path pluginsPath) {
-		this.pluginsPath = Optional.of(pluginsPath);
+	private PluginConfig(@Nullable Path pluginsPath, String[] alwaysParentFirstPatterns) {
```

Comment:
You are passing `Nullable` just to wrap it with `Optional` one line later. I would either introduce `Optional` as a parameter or avoid `Nullable` and `Optional` arguments at all - this is usually a sign that method should either be  overloaded or someone should use builder.

Reply:
Before my changes, there were two constructors, so I kind of unified them.
They say, that idiomatically java `Optional` is not supposed to be passed into constructor or  stored as field, but the use case more in the getters. In this case it would be translated to:
```java
@Nullable
private final Path pluginsPath;
...
private PluginConfig(@Nullable Path pluginsPath, String[] alwaysParentFirstPatterns) {
...
}

public Optional<Path> getPluginsPath() {
	return Optional.ofNullable(pluginsPath);
}
...
```

What do you think about such variant?
Introducing a builder now, for only two parameters, is an overkill, imo.

Reply:
> What do you think about such variant?

You know my opinion about `@Nullable` :( +1 in general, -1 in Flink as for now.

> Introducing a builder now, for only two parameters, is an overkill, imo.

Yes, it would be an overkill. That's why I would prefer lesser evil of `Optional` in the arguments